### ISSUE TITLE
Investigate using isomorphic git

### ISSUE THREAD

**Issue by habdelra** [cardstack/cardstack#567]
The nodegit dependency has been quite problematic during the life of the cardstack project. Primarily due to all the C++ bindings across the different architectures that we strive to support, and now most recently it looks like nodegit is not compatible with the more secure OpenSSH keys that are created by by default from `ssh-keygen` https://github.com/nodegit/nodegit/issues/1606.

Isomorphic git https://isomorphic-git.org/ is a pure javascript implementation of git that could alleviate all of these challenges that we've been struggling with using nodegit. Let's investigate migrating to using isomorphic git.

**Comment by ef4** [comment]
[Atwood's Law](https://blog.codinghorror.com/the-principle-of-least-power/) strikes again.

**Comment by warrick-eosny** [comment]
This is blocking windows users as nodegit seems to have problems under WSL. What I'm saying is that windows users are currently unable to complete the 1st step of the quickstart - https://docs.cardstack.com/release/getting-started/

**Comment by jenweber** [comment]
Got another report today that this is causing issues for more devs. We should bump this up in priority now that we have some more tutorials out.

**Comment by jenweber** [comment]
Note to self, when we start using isomorphic-git, we should test on Windows and see if the windows-build-tools are still necessary, and also revise the README for `@cardstack/git` and the quickstart tutorial.

**Comment by jenweber** [comment]
APIs

| nodegit | nodegit description | isomorphic-git | iso description
|---------|----------------|--------------|-----------------|
| [`Repository.open(path)`](https://www.nodegit.org/api/repository/#open) | Promise that opens a repo, use it in .then |  |  | 
| [`Repository.init(path, is_bare)`](https://www.nodegit.org/api/repository/#init)  | Promise that inits repo, use it in .then. is_bare is bool |  |  | 
| [`Branch.lookup(repo, branch_name, branch_type)`](https://www.nodegit.org/api/branch/#lookup) | promise resolves with a Reference |  |  | 
| [`Branch.create(repo, branch_name, target, force)`](https://www.nodegit.org/api/branch/#create) | creates a branch, use it in .then |  |  | 
| [`Branch.BRANCH.LOCAL`](https://www.nodegit.org/api/branch/#BRANCH) | enum with value of 1 |  |  | 
| [`Reference.nameToId(repo, name)`](https://www.nodegit.org/api/reference/#nameToId) | returns oid |  |  | 
| `Commit` |  |  |  | 
| `Merge` |  |  |  | 
| `Signature` |  |  |  | 
| `Tree` |  |  |  | 
| `setThreadSafetyStatus` |  |  |  | 
| `TreeEntry` |  |  |  | 
| `TreeBuilder` |  |  |  | 
| `Blob` |  |  |  | 
| `Cred` |  |  |  | 
| `Remote` |  |  |  | 
| `Reset` |  |  |  |

**Comment by jenweber** [comment]
After trying this refactor from a few different angles, I believe that removing nodegit from the git package is blocked and requires a rewrite. I would like to walk through this with someone else to validate.

Here are the issues I know about:
- iso does not support non-fast-forward merges. [Source](https://isomorphic-git.org/docs/en/merge). This is a blocker, because we we need this in cases where somebody else made a change in between the time you fetched a model and the time you saved it back with changes
- iso does not have a concept of tree-based operations. We could get around this by constructing our own object or filesystem in memory that tracks/mimics tree operations, then writes them to git.
- nodegit can target different references than what is currently checked out. iso follows a more human-centric flow where you cannot target arbitrary references. See [iso's commit](https://isomorphic-git.org/docs/en/commit) for example, versus [nodegit](https://www.nodegit.org/api/commit/). As far as I can tell, we rely on nodegit's fancy commit features. I tried replacing the commit itself in isogit, and returning nodegit references, and that did not work.
- In our codebase, instances of nodegit classes get passed around in a nested/chained way, and they contain state and references. If we want to use iso, we need a way to track state ourselves. This also makes piecemeal replacement very difficult.

TLDR not having non-fast-forward-merge is a hard blocker, and the remaining points mean that a rewrite is probably more feasible than a refactor.

**Comment by habdelra** [comment]
another option would be to get William Hamiton's assistance to implement non-fast-forward merges in isomorphic git. he's helped us out in the past for other things, so it's within the realm of possibility.

**Comment by ef4** [comment]
Agreed on non-fast-forward merges. I can understand why it isn't implemented yet, it's hard and there are judgement calls to make.

> iso does not have a concept of tree-based operations.

Most likely it does, but not in the public API. It may be worth poking around to confirm, but I would expect that it must necessarily do those things internally.

**Comment by wmhilton** [comment]
Hello! Yup, non-fast-forward-merges is indeed in the realm of possibility. In fact it is something that [Stoplight](https://stoplight.io) would like as well, although it hasn't ever made it onto the Kanban board. I'll make a ticket for it and see if I can't get it into the pipeline for later this month?

---

as for the tree-based operations... you can do a lot with [readObject](https://isomorphic-git.org/docs/en/readObject) and [writeObject](https://isomorphic-git.org/docs/en/writeObject). For some reason I _still_ don't have great docs on the return types of `readObject`... but it's fairly well "documented" in the TypeScript definitions here: [index.d.ts](https://github.com/isomorphic-git/isomorphic-git/blob/master/src/index.d.ts#L15-L74). Instead of using `TreeBuilder` with a you can use an object like this:

```js
await writeObject({
  dir,
  type: 'tree',
  object: {
    "entries": [
        {
          "mode": "040000",
          "path": ".circleci",
          "oid": "7d883fa42931ac76134bcf45e73de9e3d20ee74c",
          "type": "tree"
        },
        {
          "mode": "100644",
          "path": ".codeclimate.yml",
          "oid": "7a3fc521722ef9d5509d33f409555396dd2bc01c",
          "type": "blob"
        },
     ]
  }
});
```

**Comment by wmhilton** [comment]
I might be able to help y'all out directly too depending on my capacity.

**Comment by jenweber** [comment]
Hi @wmhilton thanks for chiming in and thanks for your work on this library! I’m trying out those APIs and putting together a plan.

**Comment by wmhilton** [comment]
I'll be at Epcot this weekend (perks of living in Florida!) but definitely have some availability next week during the evenings and next weekend.

**Comment by jenweber** [comment]
Hi @wmhilton just to follow up, I am waiting on one of the engineers to get back from vacation, so they can review the plan, and then we'll be in touch early next week!

**Comment by wmhilton** [comment]
As a follow up on my end, implementing basic `merge` is on the Kanban board for this sprint at Stoplight. I should have basic merges working by the end of the week.

**Comment by wmhilton** [comment]
As of [v0.65.0](https://github.com/isomorphic-git/isomorphic-git/releases/tag/v0.65.0) isomorphic-git can do full-fledged non-fast-forward merges! I'm not shouting about it from the rooftops yet, but by this weekend I'll have updated the documentation and be thinking about writing a blog post about this new feature.

If it would help, I can go through the [APIs](https://github.com/cardstack/cardstack/issues/567#issuecomment-516135223) table you made and fill in the isomorphic-git column.

**Comment by jenweber** [comment]
Hi @wmhilton that's awesome news! Thank you!

You can ignore the table - I figured things out halfway through and don't need it anymore.

The last piece I think we are missing is a public API for directly manipulating a repo (like a “bare repo” use case). We would use it to write blob and tree objects, create commits (including a merge commit), and create/update a branch. From my conversations with my colleagues, it sounds like that sort of operation may already exist behind the scenes in isomorphic-git. Do you have any ideas on this? I looked at `writeObject` and it didn't seem to directly fit our use case.

cc @ef4 who helped with figuring this out.

**Comment by wmhilton** [comment]
> a public API for directly manipulating a repo (like a “bare repo” use case). We would use it to write blob and tree objects, create commits (including a merge commit), and create/update a branch.

There's a lot of that in place now! blobs and trees aren't too hard create with `writeObject` (although I could add a `writeBlob` and `writeTree` to make it even easier) and [`commit`](https://isomorphic-git.org/docs/en/commit) now lets you override all the fields including `parent` and `tree` and combined with `noUpdateBranch: true` lets you create "branch-free, working-dir-free" commit objects. I need to update the documentation for `merge` this weekend but it should also have similar features now.
